cpp/include/cuml/experimental/fil/detail/infer_kernel/gpu.cuh

```diff
@@ -0,0 +1,255 @@
+
+  using output_t = std::conditional_t<
+    has_vector_leaves,
+    std::remove_pointer_t<vector_output_t>,
+    typename node_t::threshold_type
+  >;
+
+  using io_t = typename forest_t::io_type;
+
+  for (
+    auto i=blockIdx.x * chunk_size;
+    i < row_count;
+    i += chunk_size * gridDim.x
+  ) {
+
+    shared_mem.clear();
+    auto* output_workspace = shared_mem.fill<output_t>(output_workspace_size);
+
+    // Handle as many rows as requested per loop or as many rows as are left to
+    // process
+    auto rows_in_this_iteration = min(chunk_size, row_count - i);
+
+    auto* input_data = shared_mem.copy(
+      input + i * col_count,
+      rows_in_this_iteration,
+      col_count
+    );
+
+    auto task_count = chunk_size * forest.tree_count();
+
+    auto num_grove = raft_proto::ceildiv(
+      min(index_type(blockDim.x), task_count),
+      chunk_size
+    );
+
+    // Note that this sync is safe because every thread in the block will agree
+    // on whether or not a sync is required
+    shared_mem.sync();
+
+    // Every thread must iterate the same number of times in order to avoid a
+    // deadlock on __syncthreads, so we round the task_count up to the next
+    // multiple of the number of threads in this block. We then only perform
+    // work within the loop if the task_index is below the actual task_count.
+    auto const task_count_rounded_up = blockDim.x * raft_proto::ceildiv(task_count, blockDim.x);
+
+    // Infer on each tree and row
+    for (
+      auto task_index = threadIdx.x;
+      task_index < task_count_rounded_up;
+      task_index += blockDim.x
+    ) {
+      auto row_index = task_index % chunk_size;
+      auto real_task = task_index < task_count && row_index < rows_in_this_iteration;
+      row_index *= real_task;
+      auto tree_index = task_index * real_task / chunk_size;
+      auto grove_index = threadIdx.x / chunk_size;
+
+      auto tree_output = std::conditional_t<
+        has_vector_leaves, typename node_t::index_type, typename node_t::threshold_type
+      >{};
+      if constexpr (has_nonlocal_categories) {
+        tree_output = evaluate_tree<has_vector_leaves>(
+          forest.get_tree_root(tree_index),
+          input_data + row_index * col_count,
+          categorical_data
+        );
+      } else {
+        tree_output = evaluate_tree<has_vector_leaves, has_categorical_nodes>(
+          forest.get_tree_root(tree_index),
+          input_data + row_index * col_count
+        );
+      }
+
+      if constexpr (has_vector_leaves) {
+        for (
+          auto class_index=index_type{};
+          class_index < num_outputs;
+          ++class_index
+        ) {
+          if (real_task) {
+            output_workspace[
+              row_index * num_outputs * num_grove
+              + class_index * num_grove
+              + grove_index
+            ] += vector_output_p[
+              tree_output * num_outputs + class_index
+            ];
```

Comment:
Let's consider using `raft::mdspan` in the future, so that we can abstract away the computation of indexes.

Reply:
It wasn't available when this was written, but I agree wholeheartedly.